File: zai/scripts/vision.js
#!/usr/bin/env node
/**
 * Z.AI Vision Helper Script
 * Usage: node vision.js [options] <image_url> "Question about image"
 * 
 * Options:
 *   --model=<model>      Model to use (default: glm-4.6v)
 *   --think              Enable thinking mode
 *   --stream             Stream the response
 */

const OpenAI = require('openai');

// Parse arguments
const args = process.argv.slice(2);
let imageUrl = '';
let question = '';
let model = 'glm-4.6v';
let think = false;
let stream = false;

for (const arg of args) {
    if (arg.startsWith('--model=')) {
        model = arg.split('=')[1];
    } else if (arg === '--think') {
        think = true;
    } else if (arg === '--stream') {
        stream = true;
    } else if (!arg.startsWith('--')) {
        if (!imageUrl) {
            imageUrl = arg;
        } else {
            question = arg;
        }
    }
}

if (!imageUrl || !question) {
    console.error('Usage: node vision.js [options] <image_url> "Question about image"');
    console.error('');
    console.error('Options:');
    console.error('  --model=<model>      Model to use (default: glm-4.6v)');
    console.error('  --think              Enable thinking mode');
    console.error('  --stream             Stream the response');
    console.error('');
    console.error('Example:');
    console.error('  node vision.js https://example.com/image.png "What is in this image?"');
    process.exit(1);
}

async function main() {
    const apiKey = process.env.ZAI_API_KEY;
    if (!apiKey) {
        console.error('Error: ZAI_API_KEY environment variable not set');
        process.exit(1);
    }
    
    const client = new OpenAI({
        apiKey,
        baseURL: 'https://api.z.ai/api/paas/v4/'
    });
    
    const messages = [
        {
            role: 'user',
            content: [
                { type: 'image_url', image_url: { url: imageUrl } },
                { type: 'text', text: question }
            ]
        }
    ];
    
    const requestOpts = {
        model,
        messages,
    };
    
    if (think) {
        requestOpts.extra_body = { thinking: { type: 'enabled' } };
    }
    
    if (stream) {
        requestOpts.stream = true;
        const streamResponse = await client.chat.completions.create(requestOpts);
        
        for await (const chunk of streamResponse) {
            const delta = chunk.choices[0].delta;
            
            if (delta.reasoning_content) {
                process.stderr.write(`\x1b[90m${delta.reasoning_content}\x1b[0m`);
            }
            
            if (delta.content) {
                process.stdout.write(delta.content);
            }
        }
        console.log();
    } else {
        const response = await client.chat.completions.create(requestOpts);
        console.log(response.choices[0].message.content);
    }
}

main().catch(err => {
    console.error('Error:', err.message);
    process.exit(1);
});

File: qwen/scripts/vision.py
#!/usr/bin/env python3
"""
Qwen Vision Script
Usage: python vision.py "Describe this image" --image path/to/image.jpg
"""

import argparse
import base64
import os
import sys
from openAI import OpenAI


def encode_image(image_path):
    """Encode image to base64"""
    with open(image_path, 'rb') as f:
        return base64.b64encode(f.read()).decode('utf-8')


def main():
    parser = argparse.ArgumentParser(description='Analyze images with Qwen-VL')
    parser.add_argument('prompt', help='Your question about the image')
    parser.add_argument('--image', required=True, help='Path to image file')
    parser.add_argument('--model', default='qwen2.5-vl-7b-instruct', help='Model name')
    parser.add_argument('--url', help='Image URL instead of file')

    args = parser.parse_args()

    # Check for API key
    api_key = os.environ.get('DASHSCOPE_API_KEY') or os.environ.get('OPENAI_API_KEY')
    if not api_key:
        print('Error: Set DASHSCOPE_API_KEY or OPENAI_API_KEY environment variable')
        sys.exit(1)

    # Initialize client
    client = OpenAI(
        api_key=api_key,
        base_url=os.environ.get('OPENAI_BASE_URL', 'https://dashscope.aliyuncs.com/compatible-mode/v1')
    )

    # Prepare image content
    if args.url:
        image_url = args.url
    else:
        if not os.path.exists(args.image):
            print(f'Error: Image file not found: {args.image}')
            sys.exit(1)

        # Detect image type
        ext = os.path.splitext(args.image)[1].lower()
        mime_types = {
            '.jpg': 'image/jpeg',
            '.jpeg': 'image/jpeg',
            '.png': 'image/png',
            '.gif': 'image/gif',
            '.webp': 'image/webp'
        }
        mime_type = mime_types.get(ext, 'image/jpeg')

        base64_image = encode_image(args.image)
        image_url = f'data:{mime_type};base64,{base64_image}'

    messages = [
        {
            'role': 'user',
            'content': [
                {'type': 'image_url', 'image_url': {'url': image_url}},
                {'type': 'text', 'text': args.prompt}
            ]
        }
    ]

    try:
        response = client.chat.completions.create(
            model=args.model,
            messages=messages
        )

        print(response.choices[0].message.content)

    except Exception as e:
        print(f'Error: {e}', file=sys.stderr)
        sys.exit(1)


if __name__ == '__main__':
    main()
